src/main/scala/is/hail/variant/VariantSampleMatrix.scala

```diff
@@ -1615,6 +1615,52 @@ class VariantSampleMatrix[RPK, RK, T >: Null](val hc: HailContext, val metadata:
     (t, f2)
   }
 
+  def reorderSamples(newIds: java.util.ArrayList[Annotation]): VariantSampleMatrix[RPK, RK, T] =
+    reorderSamples(newIds.asScala.toArray)
+
+  def reorderSamples(newIds: Array[Annotation]): VariantSampleMatrix[RPK, RK, T] = {
+    requireUniqueSamples("reorder_samples")
+
+    val oldOrder = sampleIds.zipWithIndex.toMap
+    val newOrder = newIds.zipWithIndex.toMap
+
+    val mapping = oldOrder.outerJoin(newOrder).map { case (s, (oldIdx, newIdx)) =>
+      (oldIdx, newIdx) match {
+        case (Some(i), Some(j)) => (i, j)
+        case (None, Some(j)) => fatal(s"`$s' is not in dataset.")
+        case (Some(i), None) => fatal(s"dataset sample id `$s' not found in new id list.")
+        case (None, None) => fatal(s"This case should never happen!")
+      }
+    }
+
+    val newIndices = sampleIds.indices.map(mapping)
+
+    val newAnnotations = new Array[Annotation](nSamples)
+    sampleAnnotations.zipWithIndex.foreach { case (sa, idx) =>
+      newAnnotations(newIndices(idx)) = sa
+    }
+
+    val nSamplesLocal = nSamples
+    val localtct = tct
+
+    val reorderedRdd = rdd.mapPartitions({ it =>
+      it.map { case (v, (va, gs)) =>
+        val reorderedGs = localtct.newArray(nSamplesLocal)
```

Comment:
From my profiling adventures, it seems like `ClassTag.newArray` is an especially slow operation. I'm not sure why. Doesn't this just create an array of anyref?

Reply:
I think if the `val localtct` is changed to `implicit val localtct` then the `new Array[T](nSamplesLocal)` syntax should work. I'm surprised if that is any different, performance wise.

Reply:
@tpoterba I couldn't change this without getting either a Serialization error or a compilation error with what Dan suggested.